projects/packages/backup/actions.php

```diff
@@ -23,7 +25,8 @@
 }
 
 // Clean up expired Helper Scripts from a scheduled event.
-$add_action( 'jetpack_backup_cleanup_helper_scripts', array( 'Automattic\\Jetpack\\Backup\\Helper_Script_Manager', 'cleanup_expired_helper_scripts' ) );
+$helper_script_manager = new Helper_Script_Manager();
```

Comment:
Instantiating the class here is probably not a good idea. It'll probably force auto-loading the class even if a later plugin is going to provide a later version.

Reply:
Could you please elaborate?

I've remade that class into not being a static one to make it more testable, and the constructor doesn't do much of anything at all:

https://github.com/Automattic/jetpack/blob/883a2c962e2348d47507fd8ad3b8f4566cb179d8/projects/packages/backup-helper-script-manager/src/class-helper-script-manager.php#L76-L84

so I figure that it's cheap to just initialize an instance here.

Reply:
The point isn't the instantiating the class, it's that instantiating the class first triggers the autoloader to _load_ the class file.

The way things normally work with the Jetpack Autoloader is something like this:

* Plugin A loads, registers its copy of the Jetpack Autoloader.
* Plugin B loads, registers its copy of the Jetpack Autoloader.
* Later on, something tries to autoload the class (e.g. the hook fires and calls the static method), and the Jetpack Autoloader chooses the latest version of the class from all registered versions.

If you have this instantiation here, it instead could happen like this.

* Plugin A loads, registers its copy of the Jetpack Autoloader.
* Plugin A then instantiates `Helper_Script_Manager`, triggering loading of its copy of the class file.
* Plugin B loads, registers its copy of the Jetpack Autoloader.
* Plugin B then instantiates the `Helper_Script_Manager` manager class, and gets Plugin A's version even if B's is newer because A already triggered the loading.

Reply:
Hmm.

What would be the way to ensure that plugin A loads its own version of `Helper_Script_Manager` class, and plugin B loads its own version too - not an older one, not a newer one?

Reply:
You'd have to have each plugin have its own version of the class under different names, instead of both using `Automattic\Jetpack\Backup\Helper_Script_Manager`.

There are ways to try to rewrite Composer packages like that on installation (see some details in p6r3EZ-1YO-p2#comment-3498), but that's not something we've done in the monorepo and if we were to try to set something like that up we'd have to be careful not to have it affect everything relying on the current Jetpack Autoloader method.

Or you could have a custom script run during the build of each plugin that copies the one class file out of the package, modifies it, and writes the modified version into the plugin's directories somewhere. That could be tricky though when the new package isn't being used directly by the plugin, but instead by different packages used by the plugin.

Or you could stick with the status quo, where each use had its own copy-pasted version instead of having a package at all.

Reply:
Thanks, I'll add a static wrapper around `Helper_Script_Manager` so that autoloading gets delayed, and at least we could ensure that we always get the latest version instead of `rand()` version.

Reply:
In 2342562, I've moved the implementation of the helper script manager to `Helper_Script_Manager_Impl` (to make it still testable with custom locations etc.), and added a static `Helper_Script_Manager` wrapper that gets lazy-late-loaded in `add_action()` and such.